src/ts/state/actions.ts

```diff
@@ -127,10 +126,22 @@ export const removeAllMatches = () => ({
 });
 export type ActionRemoveAllMatches = ReturnType<typeof removeAllMatches>;
 
-export type Action<TMatch extends IMatch> =
+export const setFilterState = <TPluginState extends IPluginState>(
+  filterState: TPluginState["filterState"]
+) => ({
+  type: SET_FILTER_STATE,
+  payload: { filterState }
+});
+// tslint:disable-next-line:interface-over-type-literal
+export type ActionSetFilterState<TPluginState extends IPluginState> = {
+  type: typeof SET_FILTER_STATE;
+  payload: { filterState: TPluginState["filterState"] };
+};
```

Comment:
Am I right to guess that this can't be `ReturnType<typeof setFilterState>` as `tsc` doesn't infer the type correctly?

Reply:
You are! The function has a type parameter, so our usual ReturnType magic doesn't work here.

Reply:
I thought as much. OOI how does it differ from `setConfigValue` which also seems to have types?

Reply:
The parameter types in setConfigValue aren't meant to be passed by the caller – they're used to infer argument types at the point the function is called.

`ConfigKey extends keyof IPluginConfig` is really saying 'any string literal in the union of the string keys of IPluginConfig', and we're using that type at the point of call to infer the `ConfigValue`.

This means we can only call `setConfigValue` with key-value pairs that are members of `IPluginConfig`, making it typesafe for consumers.

To demonstrate, you can see further down on L148 – we're not passing the` IPluginState` param into `setConfigValue`, but we do in e.g. `ActionRequestMatchesSuccess`.